api/api/models_api.go

```diff
@@ -99,3 +104,126 @@ func (c *ModelsController) GetModel(r *http.Request, vars map[string]string, bod
 
 	return Ok(model)
 }
+
+// DeleteModel delete a model given a project and model ID
+func (c *ModelsController) DeleteModel(r *http.Request, vars map[string]string, body interface{}) *Response {
+	ctx := r.Context()
+
+	projectID, err := models.ParseID(vars["project_id"])
+	if err != nil {
+		return BadRequest("Unable to parse project_id")
+	}
+	modelID, err := models.ParseID(vars["model_id"])
+	if err != nil {
+		return BadRequest("Unable to parse model_id")
+	}
+
+	_, err = c.ProjectsService.GetByID(ctx, int32(projectID))
+	if err != nil {
+		return NotFound(err.Error())
+	}
+
+	model, err := c.ModelsService.FindByID(ctx, modelID)
+	if err != nil {
+		if gorm.IsRecordNotFoundError(err) {
+			return NotFound(fmt.Sprintf("Model is not found: %s", err.Error()))
+		}
+		return InternalServerError(err.Error())
+	}
+
+	// get all model version for this model
+	versions, _, err := c.VersionsService.ListVersions(ctx, modelID, c.MonitoringConfig, service.VersionQuery{})
+	if err != nil {
+		return InternalServerError(err.Error())
+	}
+
+	var allJobInModel []*models.PredictionJob
+	var allEndpointInModel []*models.VersionEndpoint
+
+	// iterate for every version, check the active prediction job / active endpoint
+	for _, version := range versions {
+		if model.Type == "pyfunc_v2" {
+			// check active prediction job
+			// if there are any active prediction job using the model version, deletion of the model version are prohibited
+			jobs, response := c.VersionsController.checkActivePredictionJobs(ctx, model, version)
+			if response != nil {
+				return response
+			}
+
+			// save all prediction jobs for deletion process later on
+			allJobInModel = append(allJobInModel, jobs...)
+
+		} else {
+			// handle for model with type non pyfunc
+			// check active endpoints
+			// if there are any active endpoint using the model version, deletion of the model version are prohibited
+			endpoints, response := c.VersionsController.checkActiveEndpoints(ctx, model, version)
+			if response != nil {
+				return response
+			}
+
+			// save all endpoint for deletion process later on
+			allEndpointInModel = append(allEndpointInModel, endpoints...)
+
+		}
+	}
+
+	// DELETING ALL THE RELATED ENTITY
+	for _, version := range versions {
+		if model.Type == "pyfunc_v2" {
+			// DELETE PREDICTION JOBS
+			response := c.VersionsController.deleteInactivePredictionJobs(ctx, allJobInModel, model, version)
```

Comment:
This refactoring looks good, though there's a tiny issue with the logic behind the loops. When you check for active prediction jobs/endpoints, you loop through each version of the model, and append all those jobs/endpoints in the list `allJobInModel ` or `allEndpointInModel`. That means that these lists will have a mixture of jobs/endpoints from different versions, e.g. `[v1.job1, v1.job2, v2.job1, v2.job2]`.

However, when you loop through each version again to delete the jobs/endpoints, you're passing this entire list together with with 1 version at a time to the `deleteInactivePredictionJobs` or `deleteInactiveVersionEndpoints` call, i.e. you're calling `deleteInactivePredictionJobs` with version `v1` but these jobs `[v1.job1, v1.job2, v2.job1, v2.job2]`, and then with version `v2` but also with the jobs `[v1.job1, v1.job2, v2.job1, v2.job2]`. 

This might cause an error when the deletion method searches for those inexistent jobs right (that don't correspond to their versions)?

Reply:
Resolving this conversation because it has been fixed in the newer commits.